From df02452f3df069a59bc9e69c84435bf115cb6e37 Mon Sep 17 00:00:00 2001
From: Ming Lei <ming.lei@redhat.com>
Date: Fri, 23 Sep 2022 19:51:19 +0800
Subject: [PATCH] cgroup: cgroup_get_from_id() must check the looked-up kn is a
 directory
Git-commit: df02452f3df069a59bc9e69c84435bf115cb6e37
Patch-mainline: v6.0-rc7
References: git-fix
Modified-by-SEL: Yes, modified due to different context


cgroup has to be one kernfs dir, otherwise kernel panic is caused,
especially cgroup id is provide from userspace.

Reported-by: Marco Patalano <mpatalan@redhat.com>
Fixes: 6b658c4863c1 ("scsi: cgroup: Add cgroup_get_from_id()")
Cc: Muneendra <muneendra.kumar@broadcom.com>
Signed-off-by: Ming Lei <ming.lei@redhat.com>
Acked-by: Mukesh Ojha <quic_mojha@quicinc.com>
Cc: stable@vger.kernel.org # v5.14+
Signed-off-by: Tejun Heo <tj@kernel.org>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 kernel/cgroup/cgroup.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -5985,6 +5985,8 @@ struct cgroup *cgroup_get_from_id(u64 id
 	struct kernfs_node *kn;
 	struct cgroup *cgrp = NULL;
 
+	if (kernfs_type(kn) != KERNFS_DIR)
+		goto put;
 	mutex_lock(&cgroup_mutex);
 	kn = kernfs_find_and_get_node_by_id(cgrp_dfl_root.kf_root, id);
 	if (!kn)
@@ -5996,6 +5998,8 @@ struct cgroup *cgroup_get_from_id(u64 id
 	kernfs_put(kn);
 out_unlock:
 	mutex_unlock(&cgroup_mutex);
+put:
+	kernfs_put(kn);
 	return cgrp;
 }
 EXPORT_SYMBOL_GPL(cgroup_get_from_id);
